examples/virtual-dom-wip/platform/Html/Internal/Client.roc

```diff
@@ -0,0 +1,538 @@
+# replaceHandler = \rendered, index, newHandler ->
+#     { list } = List.replace rendered.handlers index (Ok newHandler)
+#     { rendered & handlers: list }
+# -------------------------------
+#   DIFF
+# -------------------------------
+diff : DiffState state, Html state -> DiffState state
+diff = \{ rendered, patches }, newNode ->
+    oldNode =
+        List.get rendered.nodes rendered.root
+        |> Result.withDefault (Ok RenderedNone)
+        |> Result.withDefault (RenderedNone)
+
+    when { oldNode, newNode } is
+        { oldNode: RenderedText oldContent, newNode: Text newContent } ->
+            if newContent != oldContent then
+                newNodes =
+                    List.set rendered.nodes rendered.root (Ok (RenderedText newContent))
+
+                {
+                    rendered: { rendered &
+                        nodes: newNodes,
+                    },
+                    patches: List.append patches (UpdateTextNode rendered.root newContent),
+                }
+            else
+                { rendered, patches }
+
+        { oldNode: RenderedElement _oldName _oldAttrs _oldChildren, newNode: Element _newName _newSize _newAttrs _newChildren } ->
+            # TODO: actual diffing! LOL This is just to get something up and running
+            replaceNode { rendered, patches } rendered.root newNode
+
+        { oldNode: RenderedNone, newNode: None } ->
+            { rendered, patches }
+
+        _ ->
+            # old node has been replaced with a totally different variant. There's no point in diffing, just replace.
+            replaceNode { rendered, patches } rendered.root newNode
+
+replaceNode : DiffState state, NodeId, Html state -> DiffState state
+replaceNode = \diffState, oldNodeId, newNode ->
+    { rendered: createRendered, patches: createPatches, id: createNodeId } =
+        createNode diffState newNode
+    preDeleteState = {
+        rendered: createRendered,
+        patches: List.append createPatches (ReplaceNode oldNodeId createNodeId),
+    }
+
+    deleteNode preDeleteState oldNodeId
+
+deleteNode : DiffState state, NodeId -> DiffState state
+deleteNode = \diffState, id ->
+    { rendered, patches } =
+        when List.get diffState.rendered.nodes id is
+            Ok node ->
+                when node is
+                    Ok (RenderedElement _ _ children) ->
+                        List.walk children diffState deleteNode
+
+                    _ -> diffState
+
+            _ -> diffState
+    newNodes = List.set rendered.nodes id (Err DeletedNode)
+    newDeletedNodeCache = List.append rendered.deletedNodeCache id
+    newPatches = List.append patches (RemoveNode id)
+
+    {
+        rendered: { rendered &
+            nodes: newNodes,
+            deletedNodeCache: newDeletedNodeCache,
+        },
+        patches: newPatches,
+    }
+
+createNode : DiffState state, Html state -> { rendered : RenderedTree state, patches : List Patch, id : NodeId }
```

Comment:
I guess this is one place where it might be kind of nice to do

```
crateNode : DiffState state, Html state -> (DiffState state)&{ patches : List Patch }
```

(you could do that with extension variables already, but requires changing the API of `DiffState`.) Nothing actionable here, just thinking out loud... cc @rtfeldman in case you find this interesting too